File: web/src/remote.ts
import { existsSync, readFileSync } from "node:fs";
import { readFile, writeFile } from "node:fs/promises";
import { join } from "node:path";

export async function remoteAdd(path: string, name: string, url: string): Promise<void> {
	const gitDir = join(path, ".git");

	if (!existsSync(gitDir)) {
		throw new Error("Not a git repository");
	}

	const configPath = join(gitDir, "config");
	let configContent = "";

	if (existsSync(configPath)) {
		configContent = await readFile(configPath, "utf-8");
	}

	const remotePattern = new RegExp(`\\[remote\\s+"${name}"\\]`);
	if (remotePattern.test(configContent)) {
		throw new Error(`Remote '${name}' already exists`);
	}

	const remoteConfig = `[remote "${name}"]
	url = ${url}
	fetch = +refs/heads/*:refs/remotes/${name}/*
`;

	configContent = configContent.trim() + "\n\n" + remoteConfig.trim() + "\n";

	await writeFile(configPath, configContent, "utf-8");
}

export function hasRemote(path: string, name: string = "origin"): boolean {
	const gitDir = join(path, ".git");

	if (!existsSync(gitDir)) {
		return false;
	}

	const configPath = join(gitDir, "config");

	if (!existsSync(configPath)) {
		return false;
	}

	try {
		const configContent = readFileSync(configPath, "utf-8");
		const remotePattern = new RegExp(`\\[remote\\s+"${name}"\\]`);
		return remotePattern.test(configContent);
	} catch {
		return false;
	}
}

export async function setRemoteUrl(path: string, name: string, url: string): Promise<void> {
	const gitDir = join(path, ".git");
	const configPath = join(gitDir, "config");

	if (!existsSync(configPath)) {
		throw new Error("Not a git repository");
	}

	const configContent = await readFile(configPath, "utf-8");
	const lines = configContent.split("\n");

	let inRemoteSection = false;
	let currentRemote = "";
	const updatedLines: string[] = [];

	for (const line of lines) {
		const trimmed = line.trim();

		const sectionMatch = trimmed.match(/^\[remote\s+"([^"]+)"\]$/);
		if (sectionMatch) {
			inRemoteSection = true;
			currentRemote = sectionMatch[1] as string;
			updatedLines.push(line);
			continue;
		}

		if (inRemoteSection && currentRemote === name) {
			if (trimmed.startsWith("url")) {
				updatedLines.push(`\turl = ${url}`);
				continue;
			}
		}

		if (trimmed.startsWith("[") && !trimmed.startsWith("[remote")) {
			inRemoteSection = false;
		}

		updatedLines.push(line);
	}

	await writeFile(configPath, updatedLines.join("\n"), "utf-8");
}

File: dotnet/Gitologist/Remote.cs
using System.Text.RegularExpressions;

namespace Gitologist;

public static class Remote
{
    public static async Task AddRemote(string path, string name, string url)
    {
        var gitDir = Path.Combine(path, ".git");

        if (!Directory.Exists(gitDir))
        {
            throw new InvalidOperationException("Not a git repository");
        }

        var configPath = Path.Combine(gitDir, "config");
        string configContent = "";

        if (File.Exists(configPath))
        {
            configContent = await File.ReadAllTextAsync(configPath);
        }

        var remotePattern = new Regex(
            $@"\[remote\s+""{name}""\]",
            RegexOptions.Multiline
        );

        if (remotePattern.IsMatch(configContent))
        {
            throw new InvalidOperationException($"Remote '{name}' already exists");
        }

        var remoteConfig =
            $"[remote \"{name}\"]\n\turl = {url}\n\tfetch = +refs/heads/*:refs/remotes/{name}/*";

        configContent = configContent.Trim() + "\n\n" + remoteConfig.Trim() + "\n";

        await File.WriteAllTextAsync(configPath, configContent);
    }

    public static bool HasRemote(string path, string name = "origin")
    {
        var gitDir = Path.Combine(path, ".git");

        if (!Directory.Exists(gitDir))
        {
            return false;
        }

        var configPath = Path.Combine(gitDir, "config");

        if (!File.Exists(configPath))
        {
            return false;
        }

        try
        {
            var configContent = File.ReadAllText(configPath);
            var remotePattern = new Regex(
                $@"\[remote\s+""{name}""\]",
                RegexOptions.Multiline
            );
            return remotePattern.IsMatch(configContent);
        }
        catch
        {
            return false;
        }
    }

    public static async Task<string?> GetRemoteUrl(string gitDir, string remoteName)
    {
        var configPath = Path.Combine(gitDir, "config");

        if (!File.Exists(configPath))
        {
            return null;
        }

        var configContent = await File.ReadAllTextAsync(configPath);
        var lines = configContent.Split('\n');

        var inRemoteSection = false;
        var currentRemote = "";

        foreach (var line in lines)
        {
            var trimmed = line.Trim();

            var sectionMatch = Regex.Match(trimmed, @"^\[remote\s+""([^""]+)""\]");
            if (sectionMatch.Success)
            {
                inRemoteSection = true;
                currentRemote = sectionMatch.Groups[1].Value;
                continue;
            }

            if (inRemoteSection && currentRemote == remoteName)
            {
                var urlMatch = Regex.Match(trimmed, @"^url\s*=\s*(.+)$");
                if (urlMatch.Success)
                {
                    return urlMatch.Groups[1].Value.Trim();
                }
            }

            if (trimmed.StartsWith("[") && !trimmed.StartsWith("[remote"))
            {
                inRemoteSection = false;
            }
        }

        return null;
    }
}
